File: semester-3/paradigms/training/list_1/tasks.ml
let rec flatten xs =
  if xs = [] then []
  else List.append (List.hd xs) (flatten (List.tl xs))

let rec count (elem, xs) =
  if xs = [] then 0
  else (if (List.hd) xs = elem then 0 else 1) + count (elem, List.tl xs)

let rec replicate (elem, count) =
  if count = 1 then [elem]
  else elem::replicate (elem, count - 1)

let rec sqrList xs =
  if xs = [] then []
  else ((List.hd xs) * List.hd(xs))::sqrList (List.tl xs)

let reverse xs =
  let rec reverse_help acc list =
    if list = [] then acc
    else reverse_help ((List.hd list)::acc) (List.tl list)
  in reverse_help [] xs

let palindrome list =
  list = reverse list

let rec listLength xs =
  if xs = [] then 0
  else 1 + listLength (List.tl xs)

File: semester-3/paradigms/training/list_1/tasks.scala
object Tasks {
    def flatten(list: List[List[Int]]): List[Int] = list match {
        case List() => List() 
        case hd :: tail => hd ++ flatten(tail)
    }

    def count[A](elem: A, list: List[A]): Int = list match {
        case List() => 0
        case hd :: tail =>
            if (hd == elem) 1 + count(elem, tail) else count(elem, tail) 
    }

    def replicate[A](elem: A, count: Int): List[A] = count match {
        case 1 => List(elem)
        case _ => elem :: replicate(elem, count - 1)
    }

    def sqrList(list: List[Int]): List[Int] = list match {
        case List() => List()
        case hd :: tail => {
            val squared = Math.pow(hd, 2).toInt;
            squared :: sqrList(tail)
        }
    }

    def palindrome[A](list: List[A]): Boolean = list == list.reverse

    def listLength[A](list: List[A]): Int = list match {
        case List() => 0
        case hd :: tail => 1 + listLength(tail)
    }
}